upstairs/src/block_req.rs

```diff
@@ -78,53 +49,58 @@ impl Drop for BlockRes {
  * When BlockOps are sent to a guest, the calling function receives a waiter
  * that it can block on. This uses a single-use channel to receive the
  * result of a particular operation, and is meant to be paired with a
- * BlockReq.
+ * sender stored in the `BlockOp`.
  */
 #[must_use]
-pub(crate) struct BlockReqWaiter {
-    recv: oneshot::Receiver<BlockReqReply>,
+pub(crate) struct BlockReqWaiter<T, E = CrucibleError> {
+    recv: oneshot::Receiver<Result<T, E>>,
 }
 
-impl BlockReqWaiter {
+impl<T, E> BlockReqWaiter<T, E> {
     /// Create associated `BlockReqWaiter`/`BlockRes` pair
-    pub fn pair() -> (BlockReqWaiter, BlockRes) {
+    pub fn pair() -> (Self, BlockRes<T, E>) {
         let (send, recv) = oneshot::channel();
         (Self { recv }, BlockRes(Some(send)))
     }
 
     /// Consume this BlockReqWaiter and wait on the message
     ///
     /// If the other side of the oneshot drops without a reply, log an error
-    pub async fn wait(self, log: &Logger) -> BlockReqReply {
+    /// and return `None`.
+    pub async fn wait_raw(self, log: &Logger) -> Option<Result<T, E>> {
```

Comment:
Should this be `pub (crate)`?

Reply:
The `BlockReqWaiter` is already `pub(crate)`, so I think `pub` and `pub(crate)` are equivalent here; I don't have a a strong preference for which to use.  `pub(crate)` is more explicit, but redundant + longer.

Reply:
I asked because `BlockRes` seemed to have a mix, despite the struct itself being `pub(crate)`.

Whatever makes sense to you works for me.

Reply:
Good point, I switched the `BlockRes::send_result` to `pub` in b63943358734ae03bbd7f6963f96ad3275d0f3cd